/*
 * Copyright 2012-2017 Johns Hopkins University
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Developed by: Roberts Group
 *               Johns Hopkins University
 *               http://biophysics.jhu.edu/roberts/
 *
 * Author(s): Elijah Roberts, Max Klein
 */
syntax = "proto2";
package lm.types;

import "robertslab/pbuf/NDArray.proto";

message BoundaryConditions {

    enum BoundaryConditionsType {
        REFLECTING              = 0;
        ABSORBING               = 1;
        PERIODIC                = 2;
        FIXED_CONCENTRATION     = 3;        // Fixed concentration at the boundary.
        FIXED_GRADIENT          = 4;        // Fixed and unique concentration at each boundary site.
        LINEAR_GRADIENT         = 5;        // A linear gradient across the boundary.
    }    
    
    optional BoundaryConditionsType global              = 1;
    optional bool axis_specific_boundaries              = 2 [default=false];
    optional BoundaryConditionsType x_plus              = 3;
    optional BoundaryConditionsType x_minus             = 4;
    optional BoundaryConditionsType y_plus              = 5;
    optional BoundaryConditionsType y_minus             = 6;
    optional BoundaryConditionsType z_plus              = 7;
    optional BoundaryConditionsType z_minus             = 8;

    // Parameters for FIXED_CONCENTRATION and FIXED_GRADIENT.
    optional int32 boundary_species                     = 10;
    optional int32 boundary_site                        = 11;

    // Parameters for FIXED_CONCENTRATION.
    optional double boundary_concentration              = 21;
    optional double boundary_concentration_x_plus       = 22;
    optional double boundary_concentration_x_minus      = 23;
    optional double boundary_concentration_y_plus       = 24;
    optional double boundary_concentration_y_minus      = 25;
    optional double boundary_concentration_z_plus       = 26;
    optional double boundary_concentration_z_minus      = 27;

    // Parameters for FIXED_GRADIENT.
    optional robertslab.pbuf.NDArray boundary_gradient  = 32;    //3D; i=x,j=y,k=z; type=double
}
